src/internal/operators/repeat.ts

```diff
@@ -60,47 +62,44 @@ import { lift } from '../util/lift';
  * , at most count times.
  * @name repeat
  */
-export function repeat<T>(count: number = -1): MonoTypeOperatorFunction<T> {
-  return (source: Observable<T>) => {
-    if (count === 0) {
-      return EMPTY;
-    } else if (count < 0) {
-      return lift(source, new RepeatOperator(-1, source));
-    } else {
-      return lift(source, new RepeatOperator(count - 1, source));
-    }
-  };
-}
 
-class RepeatOperator<T> implements Operator<T, T> {
-  constructor(private count: number,
-              private source: Observable<T>) {
-  }
-  call(subscriber: Subscriber<T>, source: any): TeardownLogic {
-    return source.subscribe(new RepeatSubscriber(subscriber, this.count, this.source));
-  }
-}
-
-/**
- * We need this JSDoc comment for affecting ESDoc.
- * @ignore
- * @extends {Ignored}
- */
-class RepeatSubscriber<T> extends Subscriber<T> {
-  constructor(destination: Subscriber<any>,
-              private count: number,
-              private source: Observable<T>) {
-    super(destination);
-  }
-  complete() {
-    if (!this.isStopped) {
-      const { source, count } = this;
-      if (count === 0) {
-        return super.complete();
-      } else if (count > -1) {
-        this.count = count - 1;
-      }
-      source.subscribe(this._unsubscribeAndRecycle());
-    }
-  }
+export function repeat<T>(count = Infinity): MonoTypeOperatorFunction<T> {
+  return (source: Observable<T>) =>
+    count <= 0
+      ? EMPTY
+      : lift(source, function (this: Subscriber<T>, source: Observable<T>) {
+          const subscriber = this;
+          let soFar = 0;
+          const subscription = new Subscription();
+          let innerSub: Subscription | null;
+          const subscribeNext = () => {
+            let syncUnsub = false;
+            innerSub = source.subscribe({
+              next: (value) => subscriber.next(value),
+              error: (err) => subscriber.error(err),
+              complete: () => {
+                if (++soFar < count) {
+                  if (innerSub) {
+                    subscription.remove(innerSub);
+                    innerSub.unsubscribe();
```

Comment:
Is there a reason `innerSub` is not nulled here?

I think that if it isn't, there will be a problem if the source is async on the first subscription and sync on a subsequent subscription - as `innerSub` won't be `null` for the subsequent completion, despite that completion being synchronous.

Reply:
nope... just didn't think of it. Also, that `remove` is superfluous, teh `unsubscribe` will do that...